File: pid_controller.c
/**
 * Project: PID Controller
 *
 * @file pid_controller.c
 *
 * @brief test code for dual output PID controller
 *
 * @author _art_of_electronics_
 * @date 2024.11.20
 * @version v1.0
 *
 * @copyright 2024 _art_of_electronics_
*/

/*****************************************************************************
                                    INCLUDES
*****************************************************************************/

#include <stdint.h>
#include "pid_controller.h"


/*****************************************************************************
                          PRIVATE DEFINES / MACROS
*****************************************************************************/

#define MIN_OUTPUT_VALUE  (0U)
#define MAX_OUTPUT_VALUE  (100U)
#define TIME_INTERVAL     (1000.0f)


/*****************************************************************************
                     PRIVATE STRUCTS / ENUMS / VARIABLES
*****************************************************************************/

static float sPidIntegralSum = 0.0f;
static float sPidPreviousError = 0.0f;
static float sPidTimeElapsed = 0.0f;


/*****************************************************************************
                         PRIVATE FUNCTION DECLARATION
*****************************************************************************/


/*****************************************************************************
                           INTERFACE IMPLEMENTATION
*****************************************************************************/

void PID_Update(float kp, float ki, float kd, float timeout,                  \
                float currentTemp, float targetTemp, uint32_t timeDelta_ms,  \
                uint8_t *outputHeater, uint8_t *outputFan)
{

  float error = targetTemp - currentTemp;
  sPidTimeElapsed += timeDelta_ms / TIME_INTERVAL;

  ///< Check timeout
  if (sPidTimeElapsed > timeout)
  {
    *outputHeater = MIN_OUTPUT_VALUE;
    *outputFan = MIN_OUTPUT_VALUE;
    return;
  }

  ///< Integral term
  sPidIntegralSum += error * (timeDelta_ms / TIME_INTERVAL);

  ///< Derivative term
  float derivative = (error - sPidPreviousError) / (timeDelta_ms / TIME_INTERVAL);

  ///< PID output
  float output = (kp * error) + (ki * sPidIntegralSum) + (kd * derivative);

  ///< Determine heater and fan output
  if (output > MIN_OUTPUT_VALUE)
  {
    *outputHeater = (output > MAX_OUTPUT_VALUE) ? MAX_OUTPUT_VALUE : (uint8_t)output;
    *outputFan = MIN_OUTPUT_VALUE;
  }
  else
  {
    *outputFan = (-output > MAX_OUTPUT_VALUE) ? MAX_OUTPUT_VALUE : (uint8_t)(-output);
    *outputHeater = MIN_OUTPUT_VALUE;
  }

  ///< Update previous error
  sPidPreviousError = error;
}


/*****************************************************************************
                        PRIVATE FUNCTION IMPLEMENTATION
*****************************************************************************/


/******************************** END OF FILE *******************************/

File: pid_controller.py
class PIDController:
    def __init__(self, kp, ki, kd, timeout):
        self.kp = kp
        self.ki = ki
        self.kd = kd
        self.integral = 0.0
        self.prev_error = 0.0
        self.timeout = timeout * 1000  # ms
        self.elapsed_time = 0  # ms

    def update(self, current_temp, target_temp, delta_time_ms):
        if self.elapsed_time >= self.timeout:
            return 0, 100  # Full cooling if timeout is reached

        error = target_temp - current_temp

        # Proportional term
        proportional = self.kp * error

        # Integral term
        self.integral += error * (delta_time_ms / 1000.0)
        integral = self.ki * self.integral

        # Derivative term
        derivative = self.kd * (error - self.prev_error) / (delta_time_ms / 1000.0)

        # PID output
        output = proportional + integral + derivative

        # Save the current error for the next cycle
        self.prev_error = error

        # Update elapsed time
        self.elapsed_time += delta_time_ms

        # Return heater and fan output
        if output > 0:
            return min(output, 100), 0  # Heater, Fan
        else:
            return 0, min(-output, 100)  # Heater, Fan
